pwiz_tools/Shared/ProteomeDb/Fasta/WebEnabledFastaImporter.cs

```diff
@@ -929,13 +1101,367 @@ private static string NullForEmpty(string strIn)
+                            }
+                            response = new ProteinSearchInfo(); // and start another
+                            id = caption = replacedBy = title = taxId = null;
+                            length = null;
+                        }
+                        break;
+                }
+            }
+        }
+
+        private static string ExtractSpeciesFromTitle(string title)
+        {
+            if (string.IsNullOrEmpty(title))
+                return null;
+            // Entrez titles can have format: "description [status] - species name" or "description [species name]"
+            // Look for bracketed text, but skip known status words
+            var start = title.LastIndexOf('[');
+            var end = title.LastIndexOf(']');
+            if (start >= 0 && end > start)
+            {
+                var bracketed = title.Substring(start + 1, end - start - 1).Trim();
+                // Skip known status words that appear before the species
+                if (bracketed.Equals(@"validated", StringComparison.OrdinalIgnoreCase) ||
+                    bracketed.Equals(@"imported", StringComparison.OrdinalIgnoreCase))
+                {
+                    // Species appears after a dash separator
+                    var dashIndex = title.IndexOf(@" - ", start, StringComparison.Ordinal);
+                    if (dashIndex > 0 && dashIndex < title.Length - 3)
+                    {
+                        // Extract everything after " - " as the species
+                        var speciesCandidate = title.Substring(dashIndex + 3).Trim();
+                        // Remove any trailing bracketed annotations if present
+                        var trailingBracket = speciesCandidate.IndexOf('[');
+                        if (trailingBracket > 0)
+                            speciesCandidate = speciesCandidate.Substring(0, trailingBracket).Trim();
+                        return speciesCandidate;
+                    }
+                    return null;
+                }
+                return bracketed;
+            }
+            return null;
+        }
+
+        private static string BuildUniprotSearchTerm(string accession, string taxId)
+        {
+            if (string.IsNullOrEmpty(accession))
+                return accession;
+            if (string.IsNullOrEmpty(taxId))
+                return accession;
+            return $@"{accession} AND (organism_id:{taxId})";
+        }
+
+        private void ProcessEntrezSummaryResult(List<string> searchterms, List<ProteinSearchInfo> responses,
+            ref ProteinSearchInfo response, string id, string caption, string replacedBy, string title,
+            string taxId, string length, string urlString)
+        {
+            // can we transfer this search to UniprotKB? Gets us the proper accession ID,
+            // and avoids downloading sequence data we already have or just don't want
+            string newSearchTerm = null;

```

Comment:
Hard question to answer - the UniProt API moves around with some frequency, I hate to disturb an implementation that already works. Also note that UniProt seems to have won the standards battle so in pratice we don't see a lot of data that use these other APIs. I feel like it ain't broke, so don't fix it.